### DISCUSSION TITLE
segmentation fault

### DISCUSSION THREAD

**Discussion by Bala-1005** [idaholab/moose#16535] [Q&A Modules: Solid mechanics] [answered]
Hello everyone,

I am combining phase field and crystal plasticity on moose and the simulation seems to be working well for some time and then stops citing segmentation fault. I followed the debugging process given on moose framework website and built a debug object file and then ran the file again. I obtained the following error.
**Time Step 5, time = 0.023, dt = 0.008
[New Thread 0x7fffc6eb2700 (LWP 56706)]
[Thread 0x7fffc6eb2700 (LWP 56706) exited]
[New Thread 0x7fffc6eb2700 (LWP 56715)]
[Thread 0x7fffc6eb2700 (LWP 56715) exited]
[New Thread 0x7fffc6eb2700 (LWP 56719)]
[Thread 0x7fffc6eb2700 (LWP 56719) exited]
[New Thread 0x7fffc6eb2700 (LWP 56747)]
/export/apps/gcc-7.4.0/include/c++/7.4.0/debug/vector:417:
Error: attempt to subscript container with out-of-bounds index 0, but
container only holds 0 elements.

Objects involved in the operation:
    sequence "this" @ 0x0x208f4c8 {
      type = std::__debug::vector<RankTwoTensorTempl<double>, std::allocator<RankTwoTensorTempl<double> > >;
    }
Program received signal SIGABRT, Aborted.
0x00007fffdeed1337 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install glibc-2.17-292.el7.x86_64 infinipath-psm-3.3-26_g604758e_open.2.el7.x86_64 libX11-1.6.7-2.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libfabric-1.7.0-1.el7.x86_64 libibumad-22.1-3.el7.x86_64 libibverbs-22.1-3.el7.x86_64 libnl3-3.2.28-4.el7.x86_64 libpsm2-11.2.78-1.el7.x86_64 librdmacm-22.1-3.el7.x86_64 libuuid-2.23.2-63.el7.x86_64 libxcb-1.13-1.el7.x86_64 numactl-libs-2.0.12-3.el7_7.1.x86_64 opensm-libs-3.3.21-2.el7.x86_64 zlib-1.2.7-18.el7.x86_64**

After typing where, I get a long list of notifications as below.

#1  0x00007fffdeed2a28 in abort () from /lib64/libc.so.6
#2  0x00007fffdfa54593 in __gnu_debug::_Error_formatter::_M_error (
    this=0x61e420 <__gnu_debug::_Error_formatter::_M_at(char const*, unsigned int)::__formatter>)
    at ../../../../../gcc-7.4.0/libstdc++-v3/src/c++11/debug.cc:1069
#3  0x00007ffff78d4c71 in std::__debug::vector<RankTwoTensorTempl<double>, std::allocator<RankTwoTensorTempl<double>                                                                                                                         > >::operator[] (this=0x208f4c8, __n=0) at /export/apps/gcc-7.4.0/include/c++/7.4.0/debug/vector:417
#4  0x00007ffff1e8f2a3 in CrystalPlasticitySlipRateGSS::calcFlowDirection (this=0x1337700, qp=2, flow_direction=...)
    at /home/gnanasbr/temp_projects/moose/modules/tensor_mechanics/src/userobjects/CrystalPlasticitySlipRateGSS.C:162
#5  0x00007ffff1fc76a2 in FiniteStrainUObasedCP::computeQpStress (this=0x149c310)
    at /home/gnanasbr/temp_projects/moose/modules/tensor_mechanics/src/materials/FiniteStrainUObasedCP.C:231
#6  0x00007ffff1fa39f2 in ComputeStressBase::computeQpProperties (this=0x149c310)
    at /home/gnanasbr/temp_projects/moose/modules/tensor_mechanics/src/materials/ComputeStressBase.C:52
.......................................................................................................................
It goes to list a number of locations in other source code files. 

I am not sure where I have gone wrong. As far as I can understand the error seems to be in the calcFlowDirection of the CrystalPlasticitySlipRateGSS user object.

How do I solve this problem?

Thanks,
Bala

**Comment by GiudGiud** [comment]
Hi Bala,

I'm not an expert in this module however the debug trace seems to point to an out of bounds access in _flow_direction.
It's initialized with a material property in the constructor:
   ``` _flow_direction(getMaterialProperty<std::vector<RankTwoTensor>>(_name + "_flow_direction")) ```

Does that give you any hint of what is the issue? Can you share a MWE with the issue?

Guillaume

**Reply by aeslaughter** [reply]
This type of error would lead me to believe that _flow_direction is not a reference when retrieved with getMaterialProperty

**Reply by Bala-1005** [reply]
Hello Guillaume and Andrew,

I understand the issue. But I am not sure why it popped up as I have used the same FiniteCrystalUOBasedCP as shown in one of the examples. And forgive my ignorance, what is an MWE?

How do I retrieve _flow_direction as a  reference?

Thanks,
Bala

**Reply by aeslaughter** [reply]
MWE = Minimum Working Example

**Reply by aeslaughter** [reply]
To use a reference you need the "&" as follows in the header:

``` 
const MaterialProperty<T> & _D;
```

**Reply by Bala-1005** [reply]
Hello Guillaume and Andrew,

Thank you for your inputs.
I have attached the input file that I used and I was able to run the file successfully for a few time steps beyond which I got the segmentation fault. 
[kks_anisotropic_plastic.txt](https://github.com/idaholab/moose/files/5730857/kks_anisotropic_plastic.txt)


As far as the flow direction goes, I found that it has been called by reference in the calcflowdirection function.
`void
CrystalPlasticitySlipRateGSS::calcFlowDirection(unsigned int qp,
                                                std::vector<RankTwoTensor> & flow_direction) const
{
  DenseVector<Real> mo(LIBMESH_DIM * _variable_size), no(LIBMESH_DIM * _variable_size);

  // Update slip direction and normal with crystal orientation
  for (unsigned int i = 0; i < _variable_size; ++i)
  {
    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
    {
      mo(i * LIBMESH_DIM + j) = 0.0;
      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
        mo(i * LIBMESH_DIM + j) =
            mo(i * LIBMESH_DIM + j) + _crysrot[qp](j, k) * _mo(i * LIBMESH_DIM + k);
    }

    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
    {
      no(i * LIBMESH_DIM + j) = 0.0;
      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
        no(i * LIBMESH_DIM + j) =
            no(i * LIBMESH_DIM + j) + _crysrot[qp](j, k) * _no(i * LIBMESH_DIM + k);
    }
  }
'
And the flow_direction variable was declared in the constructors part as 
`_flow_direction(getMaterialProperty<std::vector<RankTwoTensor>>(_name + "_flow_direction"))`

Thanks,
Bala

**Reply by GiudGiud** [reply]
How is flow direction declared in the header?

**Comment by jiangwen84** [comment, marked as answer]
@Bala-1005 We have not implemented the code in MOOSE to couple crystal plasticity with phase field fracture yet. Some external users have implemented their model in their own branch, and you might want to check with them, i.e. @ngrilli 

At INL, we have been developing a variational consistent phase field fracture model for crystal plasticity , but it is not ready yet.

**Reply by Bala-1005** [reply]
I see. Thank you for the information. But the phase-field coupling here is not for fracture but for solidification. Does it change anything?

**Reply by jiangwen84** [reply]
What model are you using for solidification? Are you using the grain tracker? Our CP needs grain orientation information, that is usually provided by users as an input. If you have new grains being generated during solidification, I do not believe our CP can handle that case at this moment.

**Reply by Bala-1005** [reply]
I am using KKS model and I am not using grain tracker. I have given grain orientation through interfaceorientationmaterial. And there are no new nucleation events. There is a small region where the phase field corresponds to solid and the phase variable evolves from this initial condition.
